modules/proxy/proxy_util.c

```diff
@@ -4765,17 +4773,12 @@ PROXY_DECLARE(apr_status_t) ap_proxy_tunnel_create(proxy_tunnel_rec **ptunnel,
     tunnel->origin->pfd->desc_type = APR_POLL_SOCKET;
     tunnel->origin->pfd->desc.s = ap_get_conn_socket(c_o);
     tunnel->origin->pfd->client_data = tunnel->origin;
+    apr_socket_timeout_get(tunnel->origin->pfd->desc.s, &origin_timeout);
+    apr_socket_opt_set(tunnel->origin->pfd->desc.s, APR_SO_NONBLOCK, 1);
 
     /* Defaults to the biggest timeout of both connections */
-    apr_socket_timeout_get(tunnel->client->pfd->desc.s, &timeout);
-    apr_socket_timeout_get(tunnel->origin->pfd->desc.s, &tunnel->timeout);
-    if (timeout >= 0 && (tunnel->timeout < 0 || tunnel->timeout < timeout)) {
-        tunnel->timeout = timeout;
-    }
-
-    /* We should be nonblocking from now on the sockets */
-    apr_socket_opt_set(tunnel->client->pfd->desc.s, APR_SO_NONBLOCK, 1);
-    apr_socket_opt_set(tunnel->origin->pfd->desc.s, APR_SO_NONBLOCK, 1);
+    tunnel->timeout = (origin_timeout >= 0 && origin_timeout > client_timeout)?
+                      origin_timeout : client_timeout;
```

Comment:
I'm wondering if we shouldn't use the smaller of the two timeouts instead (i.e. change the original behaviour). It could be easier for the user to control the lifetime/ending of ws connections by doing this..

Reply:
I was wondering about the original choice. I think it makes sense to use the smaller timeout, since this *should* tear down the tunnel, I believe.

Reply:
Yeah, makes more sense to me.

Reply:
Like this?

```
    tunnel->timeout = (client_timeout > 0 && client_timeout < origin_timeout)?
                      client_timeout : origin_timeout;
```

Reply:
The question really is what a timeout of `0` actually should do? Would the tunnel not do busy-loop polling?

Reply:
A zero timeout is probably not possible/useful but strictly speaking it would be:
```
    tunnel->timeout = (client_timeout >= 0 && client_timeout < origin_timeout ?
                       client_timeout : origin_timeout);
```
(using _my_ codestyle to please _my_ vim indent :)

Reply:
> The question really is what a timeout of `0` actually should do? Would the tunnel not exit after the first poll without an event?

I think it would timeout immediately if data on any side are not available immediately (not very useful..).

Reply:
Added as r1910704 to trunk.

Reply:
:+1: thanks Stefan!